File: leetcode-daily/2025-08-aug/2025-08-12/2025-08-12-fun-with-graphs.c
// Title: Fun with Graphs
// Date: 2025-08-12
// Created: `Tue, 12 Aug 2025 18:27:30 +0530`
// Tags: [graphs, daily-practice]

#include <memory.h>
#include <stdbool.h>
#include <stdlib.h>

/// {{
/// Problem: 547. Number of Provinces
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/number-of-provinces/
/// Topics: `depth-first-search`, `breadth-first-search`, `union-find`, `graph`
/// Timestamp: `Tue, 12 Aug 2025 18:27:30 +0530`

struct QuickUnion {
  int *entries;
  int *sizes;
  int capacity;
  int componentCount;
};
typedef struct QuickUnion *UF;

void QuickUnion_Init(UF uf, int n);
void QuickUnion_Clear(UF uf);
void QuickUnion_Union(UF uf, int p, int q);
 int QuickUnion_Find(UF uf, int p);
bool QuickUnion_Connected(UF uf, int p, int q);
 int QuickUnion_Count(UF uf);

void QuickUnion_Init(UF uf, int n) {
  uf->capacity = n;
  uf->entries = calloc(uf->capacity, sizeof(int));
  uf->sizes = calloc(uf->capacity, sizeof(int));
  uf->componentCount = n;

  for (int i = 0; i < uf->capacity; i++) {
    uf->entries[i] = i;
    uf->sizes[i] = 1;
  }
}

void QuickUnion_Clear(UF uf) {
  free(uf->entries), (uf->entries = NULL);
  free(uf->sizes), (uf->sizes = NULL);
  memset(uf, 0, sizeof(struct QuickUnion));
}

void QuickUnion_Union(UF uf, int p, int q) {
  int rootP = QuickUnion_Find(uf, p);
  int rootQ = QuickUnion_Find(uf, q);

  if (rootP == rootQ) {
    return;
  }

  int sizeP = uf->sizes[rootP];
  int sizeQ = uf->sizes[rootQ];

  if (sizeP < sizeQ) {
    uf->entries[rootP] = rootQ;
    uf->sizes[rootQ] += sizeP;
  } else {
    uf->entries[rootQ] = rootP;
    uf->sizes[rootP] += sizeQ;
  }
  uf->componentCount--;
}

int QuickUnion_Find(UF uf, int p) {
  while (uf->entries[p] != p) {
    uf->entries[p] = uf->entries[uf->entries[p]]; // path compression
    p = uf->entries[p];
  }
  return p;
}

inline bool QuickUnion_Connected(UF uf, int p, int q) {
  return QuickUnion_Find(uf, p) == QuickUnion_Find(uf, q);
}

inline int QuickUnion_Count(UF uf) {
  return uf->componentCount;
}


int findCircleNum(int **isConnected, int isConnectedSize, int *isConnectedColSize) {
  const int rows = isConnectedSize;
  const int cols = *isConnectedColSize;

  struct QuickUnion _uf, *uf = &_uf;
  QuickUnion_Init(uf, cols);

  for (int i = 0; i < rows; i++) {
    for (int j = 0; j < cols; j++) {
      if (isConnected[i][j] == 1) {
        QuickUnion_Union(uf, i, j);
      }
    }
  }

  int count = QuickUnion_Count(uf);
  QuickUnion_Clear(uf), (uf = NULL);

  return count;
}

/// }}

File: leetcode-daily/2025-08-aug/2025-08-12/2025-08-12-fun-with-graphs.js
// Title: Fun with Graphs
// Date: 2025-08-12
// Created: `Tue, 12 Aug 2025 18:46:48 +0530`
// Tags: [linked-lists, daily-practice]

/// {{
/// Problem: 547. Number of Provinces
/// Difficulty: `Medium`
/// Links: https://leetcode.com/problems/number-of-provinces/
/// Topics: `depth-first-search`, `breadth-first-search`, `union-find`, `graph`
/// Timestamp: `Tue, 12 Aug 2025 21:35:08 +0530`

class QuickUnion {
  /** @type {number[]} */
  #entries;

  /** @type {number[]} */
  #sizes;

  /** @type {number} */
  #componentCount;

  /** @param {number} n */
  constructor(n) {
    this.#entries = Array(n).fill(0);
    this.#sizes = Array(n).fill(1);
    this.#componentCount = n;

    for (let i = 0; i < n; i++) {
      this.#entries[i] = i;
    }
  }

  /**
   * @param {number} p
   * @param {number} q
   */
  union(p, q) {
    const [rootP, rootQ] = [this.find(p), this.find(q)];

    if (rootP === rootQ) {
      return;
    }

    const [sizeP, sizeQ] = [this.#sizes[rootP], this.#sizes[rootQ]];

    if (sizeP < sizeQ) {
      this.#entries[rootP] = rootQ;
      this.#sizes[rootP] += this.#sizes[rootQ];
    } else {
      this.#entries[rootQ] = rootP;
      this.#sizes[rootQ] += this.#sizes[rootP];
    }
    this.#componentCount--;
  }

  /** @param {number} p */
  find(p) {
    while (this.#entries[p] != p) {
      this.#entries[p] = this.#entries[this.#entries[p]];
      p = this.#entries[p];
    }
    return p;
  }

  count() {
    return this.#componentCount;
  }
}

/**
 * @param {number[][]} isConnected
 * @return {number}
 */
function findCircleNum(isConnected) {
  const [rows, cols] = [isConnected.length, isConnected[0].length];
  const uf = new QuickUnion(rows);

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      if (isConnected[i][j] === 1) {
        uf.union(i, j);
      }
    }
  }

  return uf.count();
}

/// }}

